crates/biome_cli/src/reporter/gitlab.rs

```diff
@@ -0,0 +1,243 @@
+        Some(category.name().to_string())
+    }
+
+    /// Extracts the source code generating the diagnostic.
+    fn source_code(&self, value: &biome_diagnostics::error::Error) -> Option<String> {
+        let location = value.location();
+        let source_code = location.source_code?;
+        let span = location.span?;
+
+        Some(source_code.text[span].to_string())
+    }
+
+    /// Generates a fingerprint for a diagnostic.
+    fn fingerprint(&self, check_name: &String, path: &String, code: &String) -> u64 {
+        let mut hasher = DefaultHasher::new();
+
+        // Including the source code in our hash leads to more stable
+        // fingerprints. If you instead rely on e.g. the line number and change
+        // the first line of a file, all of its fingerprint would change.
+        code.hash(&mut hasher);
+        check_name.hash(&mut hasher);
+        path.hash(&mut hasher);
+
+        hasher.finish()
+    }
+
+    /// Enforces uniqueness of generated fingerprints in the context of a
+    /// single report.
+    fn ensure_fingerprint_uniqueness(&mut self, fingerprint: u64, salt: u64) -> u64 {
+        let mut current = fingerprint;
+        while self.fingerprints.contains(&current) {
+            let mut hasher = DefaultHasher::new();
+            current.hash(&mut hasher);
+            salt.hash(&mut hasher);
+            current = hasher.finish();
+        }
+
+        self.fingerprints.insert(current);
+        current
+    }
+
+    /// GitLab only cares about paths relative to the repository root.
+    /// This function attempts to relativize the absolute path from the
+    /// diagnostic, while falling back to the originally reported one.
+    fn path(&self, value: &biome_diagnostics::error::Error) -> Option<String> {
+        let path = match value.location().resource? {
+            biome_diagnostics::Resource::Argv => None,
+            biome_diagnostics::Resource::Memory => None,
+            biome_diagnostics::Resource::File(path) => Some(path.to_string()),
+        }?;
+
+        let Some(root) = &self.repository_root else {
+            return Some(path);
+        };
+
+        let Ok(resolved) = canonicalize(Path::new(path.as_str())) else {
+            return Some(path);
+        };
+
+        let Ok(relativized) = resolved.strip_prefix(root) else {
+            return Some(path);
+        };
+
+        Some(relativized.to_str().unwrap_or(path.as_str()).to_string())
+    }
+
+    /// Extracts the line number from the diagnostic.
+    fn line(&self, value: &biome_diagnostics::error::Error) -> Option<u8> {
+        let location = value.location();
+        let buf = LineIndexBuf::from_source_text(location.source_code?.text);
+        let diagnostic_offset = location.span?.start();
+
+        buf.iter()
+            .enumerate()
+            .find(|(_, line_offset)| **line_offset >= diagnostic_offset)
+            .map(|(line_number, _)| u8::try_from(line_number).unwrap())
```

Comment:
Why do you use `u8` instead of `u32`? Also, don't use `unwrap`, because it's a method that panics. We want to use methods that don't panic